CARDS-1812: Computing form completion status flags overwrites any oth…

sofi2002sofi opened the pull request:
…er flags set on the form

sofi2002sofi committed:
CARDS-1812: Computing form completion status flags overwrites any other flags set on the form

sdumitriu reviewed (commented):
Code looks good, didn't test yet.

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -107,11 +109,16 @@ public void propertyChanged(final PropertyState before, final PropertyState afte
             final int numAnswers = iterableLength(nodeAnswers);
-            final List<String> statusFlags = new ArrayList<>();
+            final Set<String> statusFlags = new TreeSet<>();
+            if (before != null) {
+                before.getValue(Type.STRINGS).forEach(statusFlags::add);
+            }
```
Ah, this has two bugs:

- the `before` state is the old value, not the one you're trying to set, which means that trying to set new flags will always go back to the previous state, plus or minus INVALID/INCOMPLETE, but without any of the new values; `after` should be used instead, but:
- the `before` and `after` are only about the `value` property in here, not the status flags, which are only found in the `currentNodeBuilder`

This should work:

```suggestion
            this.currentNodeBuilder.getProperty(STATUS_FLAGS).getValue(Type.STRINGS).forEach(statusFlags::add);
```

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -143,11 +145,15 @@ public void propertyDeleted(final PropertyState before)
             if (PROP_VALUE.equals(before.getName())) {
-                final List<String> statusFlags = new ArrayList<>();
+                final Set<String> statusFlags = new TreeSet<>();
+                before.getValue(Type.STRINGS).forEach(statusFlags::add);
```
Here too the flags need to be taken from `currentNodeBuilder`.

sdumitriu reviewed (changes requested):
The same set initialization should be done in `summarize`.

sofi2002sofi committed:
CARDS-1812: Computing form completion status flags overwrites any other flags set on the form

change the source of getting previous statusFlags

sofi2002sofi commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -107,11 +109,14 @@ public void propertyChanged(final PropertyState before, final PropertyState afte
         if (questionNode != null && PROP_VALUE.equals(after.getName())) {
```
@sdumitriu At this stage, we expect to change only the "value" property of the question node. That is why the problem described [here](https://phenotips.atlassian.net/browse/CARDS-1812) couldn't be solved without changing these lines. Will getting either question node or form node, and just in the first case checking if it is the "value" property significantly affect other logic?

sdumitriu replied:
Flags for non-answer nodes (Form and AnswerSection) are computed in the `summarize` method. There are no changes needed in this method for fixing the form status flags, since this explicitly handles changes in the answer values.

sofi2002sofi committed:
CARDS-1812: Computing form completion status flags overwrites any other flags set on the form

fix summarize method

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -107,11 +109,14 @@ public void propertyChanged(final PropertyState before, final PropertyState afte
             final int numAnswers = iterableLength(nodeAnswers);
-            final List<String> statusFlags = new ArrayList<>();
+            final Set<String> statusFlags = new TreeSet<>();
+            this.currentNodeBuilder.getProperty(STATUS_FLAGS).getValue(Type.STRINGS).forEach(statusFlags::add);
```
It is possible that the property doesn't exist, so it's safer to wrap thin in an `if (this.currentNodeBuilder.hasProperty(STATUS_FLAGS))`.

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -143,11 +143,15 @@ public void propertyDeleted(final PropertyState before)
             if (PROP_VALUE.equals(before.getName())) {
-                final List<String> statusFlags = new ArrayList<>();
+                final Set<String> statusFlags = new TreeSet<>();
+                this.currentNodeBuilder.getProperty(STATUS_FLAGS).getValue(Type.STRINGS).forEach(statusFlags::add);
```
`if` here too.

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -312,7 +322,10 @@ private void summarize() throws RepositoryException
+        if (this.currentNodeBuilder.hasProperty(STATUS_FLAGS)) {
+            this.currentNodeBuilder.getProperty(STATUS_FLAGS).getValue(Type.STRINGS).forEach(statusFlags::add);
+        }
         if (isInvalid) {
```
Bug: a form never loses its incomplete status since we always start with the previous flags. You must also add the `else remove` branches.

sofi2002sofi replied:
However, while first creating the form it does not have this property so it causes a NullPointerException

sofi2002sofi committed:
CARDS-1812: Computing form completion status flags overwrites any other flags set on the form

added a check for the existence of the property, and added the removal of the invalid and incomplete status flags in the summarize method
